jabgui/src/main/java/org/jabref/gui/walkthrough/declarative/NavigationPredicate.java

```diff
@@ -0,0 +1,229 @@
+        if (!(node instanceof ContextMenuContent)
+                && Stream.iterate(node.getParent(), Objects::nonNull, Parent::getParent)
+                         .noneMatch(ContextMenuContent.class::isInstance)) {
+            return Optional.empty();
+        }
+
+        return Window.getWindows().stream()
+                     .map(Window::getScene)
+                     .filter(Objects::nonNull)
+                     .map(scene -> scene.lookup(".mainMenu"))
+                     .filter(MainMenu.class::isInstance)
+                     .map(MainMenu.class::cast)
+                     .flatMap(menu -> menu.getMenus().stream())
+                     .flatMap(topLevelMenu -> topLevelMenu.getItems().stream())
+                     .filter(menuItem -> Optional.ofNullable(menuItem.getGraphic())
+                                                 .map(graphic -> graphic.equals(node)
+                                                         || Stream.iterate(graphic, Objects::nonNull, Node::getParent)
+                                                                  .anyMatch(contextMenu -> contextMenu.equals(node)))
+                                                 .orElse(false))
+                     .findFirst();
+    }
+
+    static <T extends Event> EventHandler<T> decorate(
+            Runnable beforeNavigate,
+            EventHandler<? super T> originalHandler,
+            Runnable onNavigate) {
+        return event -> navigate(beforeNavigate, originalHandler, event, onNavigate);
+    }
+
+    /// Navigates to the target node by consuming the event.
+    /// 1. Execute beforeNavigate, which typically used to stop node polling and prevent auto-fallback
+    /// navigation during the handler execution.
+    /// 2. Execute the original event handler if it exists.
+    /// 3. Execute the onNavigate runnable, when
+    ///    - The original handler has completed execution, or
+    ///    - The handler has timed out after HANDLER_TIMEOUT_MS milliseconds.
+    ///    - A new window has been opened, or an existing window has been closed.
+    ///
+    ///    Those conditions ensure that we will still navigate if original handler is blocking (e.g., showing a dialog,
+    ///    and we are highlighting a node in the dialog) and is still responsive (i.e., we will navigate after a certain
+    ///    amount of time, or a new window has been opened, and we can start polling the new window).
+    ///
+    /// @param <T> the type of the event
+    /// @param beforeNavigate a runnable to execute before navigation
+    /// @param originalHandler the original event handler to execute
+    /// @param event the event to navigate
+    static <T extends Event> void navigate(
+            Runnable beforeNavigate,
+            EventHandler<? super T> originalHandler,
+            T event,
+            Runnable onNavigate) {
+
+        event.consume();
+        beforeNavigate.run();
+
+        CompletableFuture<Void> handlerFuture = new CompletableFuture<>();
```

Comment:
This variable is never read? -> always handlerFtuure.complete(null) passed. What is the intention?